### PR TITLE
CSM Timedelta Update

### PR DESCRIPTION
**Pull request by vhirtham**
## Checks
- [x] updated CHANGELOG.md
- [x] updated tests
- [x] updated doc/

### CODE REVIEW

File: weldx/transformations.py [hunk 1]
```diff
@@ -1237,6 +1239,8 @@ def __init__(
         if coordinate_system_manager_name is None:
             coordinate_system_manager_name = self._generate_default_name()
         self._name = coordinate_system_manager_name
+        self._reference_time = reference_time
+        self._lcs_use_ref_time = None  # only relevant if CSM has no reference time
```
Comment (vhirtham) [inline weldx/transformations.py:1243]: replace with property

File: weldx/transformations.py [hunk 2]
```diff
@@ -1609,6 +1683,20 @@ def add_cs(
...
+            )
+            if (
+                self._number_of_time_dependent_lcs > 0
+                and current_lcs_has_ref_time != self._has_lcs_with_time_ref
+            ):
+                raise Exception(
+                    "Inconsistent usage of reference times! If you didn't specify a "
+                    "reference time for the CoordinateSystemManager, either all or "
+                    "none of the added coordinate systems must have a reference time."
+                )
```
Comment (CagtayFabry) [inline weldx/transformations.py:1698]: Check combine `if`

add `CSM.uses_absolute_times`

File: weldx/transformations.py [hunk 3]
```diff
@@ -2108,7 +2199,8 @@ def get_local_coordinate_system(
         self,
         coordinate_system_name: str,
         reference_system_name: Union[str, None] = None,
-        time: Union[pd.DatetimeIndex, List, str, None] = None,
+        time: Union[pd.TimedeltaIndex, pd.DatetimeIndex, pint.Quantity, str] = None,
+        time_ref: pd.Timestamp = None,
     ) -> LocalCoordinateSystem:
```
Comment (CagtayFabry) [inline weldx/transformations.py:2204]: refactor get_cs

File: weldx/transformations.py [hunk 4]
```diff
@@ -2183,25 +2343,39 @@ def get_local_coordinate_system(
...
+        lcs_result = LocalCoordinateSystem()
+        for edge in path_edges:
+            lcs = self.graph.edges[edge]["lcs"]
+            if lcs.time is not None:
+                if lcs.reference_time is None and self.reference_time is not None:
+                    lcs = deepcopy(lcs)
+                    lcs.reset_reference_time(self.reference_time)
+                lcs = lcs.interp_time(time_interp, time_ref_interp)
+            lcs_result += lcs
+        return lcs_result
```
Comment (CagtayFabry) [inline weldx/transformations.py:2378]: check simplification

File: weldx/transformations.py [hunk 5]
```diff
@@ -2494,31 +2680,74 @@ def remove_subsystems(self):
...
+            if lcs.time is None:
+                continue
+
+            if return_time_delta:
+                time_lcs = lcs.time
+            elif lcs.reference_time is None:
+                time_lcs = self.reference_time + lcs.time
+            else:
+                time_lcs = lcs.reference_time + lcs.time
+
```
Comment (CagtayFabry) [inline weldx/transformations.py:2746]:
```python
time_list = [lcs.time for lcs in CSM.list_lcs if lcs.time is not None]
if CSM.ref_time is not None:
    time_list  = [t+CSM.ref_time if isinstance(t,pd.TimedeltaIndex) else t for t in time_list ]
```

File: weldx/asdf/schemas/weldx.bam.de/weldx/core/transformations/coordinate_system_hierarchy-1.0.0.yaml [hunk 6]
```diff
@@ -42,7 +47,7 @@ properties:
     items:
       $ref: "tag:weldx.bam.de:weldx/core/transformations/coordinate_transformation-1.0.0"
 
-propertyOrder: [name, root_system_name, subsystems, subsystem_data,  coordinate_systems]
+propertyOrder: [name, root_system_name, time_ref, subsystems, subsystem_data,  coordinate_systems]
```
Comment (CagtayFabry) [inline weldx/asdf/schemas/weldx.bam.de/weldx/core/transformations/coordinate_system_hierarchy-1.0.0.yaml:50]:
```suggestion
propertyOrder: [name, root_system_name, reference_time, subsystems, subsystem_data,  coordinate_systems]
```
To match the property name?

File: weldx/asdf/schemas/weldx.bam.de/weldx/core/transformations/local_coordinate_system-1.0.0.yaml [hunk 7]
```diff
@@ -35,7 +35,7 @@ wx_shape:
     data: [(t), 3, 3]
 
 required: [orientations, coordinates]
-propertyOrder: [reference_time, time, orientations, coordinates]
+propertyOrder: [time_ref, time, orientations, coordinates]
```
Comment (CagtayFabry) [inline weldx/asdf/schemas/weldx.bam.de/weldx/core/transformations/local_coordinate_system-1.0.0.yaml:38]:
```suggestion
propertyOrder: [reference_time, time, orientations, coordinates]
```
see property

File: tests/test_transformations.py [hunk 8]
```diff
@@ -1812,11 +1750,79 @@ def test_add_coordinate_system(
...
+    @pytest.mark.parametrize(
+        "has_timestamp_csm, has_timestamp_lcs_1, has_timestamp_lcs_2, should_raise",
+        [
+            (True, False, False, False),
+            (True, True, False, False),
+            (True, False, True, False),
+            (True, True, True, False),
+            (False, False, False, False),
+            (False, True, False, True),
+            (False, False, True, True),
```
Comment (CagtayFabry) [inline tests/test_transformations.py:1771]:
```suggestion
            (False, True, False, Exception),
            (False, False, True, Exception),
```
I like this setup, maybe explicitly list the exception here directly for more flexibility (see below)

File: tests/test_transformations.py [hunk 9]
```diff
@@ -1812,11 +1750,79 @@ def test_add_coordinate_system(
...
+        lcs_2 = tf.LocalCoordinateSystem(
+            coordinates=[[1, 5, 3], [3, 5, 1]],
+            time=pd.TimedeltaIndex([0, 2]),
+            time_ref=timestamp_lcs_2,
+        )
+
+        csm.add_cs("lcs_1", "root", lcs_1)
+
+        if should_raise:
+            with pytest.raises(Exception):
```
Comment (CagtayFabry) [inline tests/test_transformations.py:1822]:
```suggestion
            with pytest.raises(should_raise):
```
test for specific exception

File: tests/test_transformations.py [hunk 10]
```diff
@@ -1959,84 +1965,949 @@ def test_delete_coordinate_system_exceptions(
...
+        exp_results :
+            A list containing the expected results of each instance comparison
+
+        """
+        # define support function
+        def create_csm_list(csm_data_list, cs_data_list, merge_data_list):
+            """Create a list of CSM instances."""
+            csm_list = []
+            for args in csm_data_list:
+                csm_list += [tf.CoordinateSystemManager(*args)]
```
Comment (CagtayFabry) [inline tests/test_transformations.py:2255]: I think this does the same?
```suggestion
            csm_list = [tf.CoordinateSystemManager(*args) for args in csm_data_list]
```

File: tests/test_transformations.py [hunk 11]
```diff
@@ -1959,84 +1965,949 @@ def test_delete_coordinate_system_exceptions(
...
+            merge_data_diff[diff[0]] = diff[1]
+
+        # create CSM instances
+        csm_list_1 = create_csm_list(csm_data, cs_data, merge_data)
+        csm_list_2 = create_csm_list(csm_data_diff, cs_data_diff, merge_data_diff)
+
+        # test
+        for i, _ in enumerate(csm_list_1):
+            assert (csm_list_1[i] == csm_list_2[i]) is exp_results[i]
+            assert (csm_list_1[i] != csm_list_2[i]) is not exp_results[i]
```
Comment (CagtayFabry) [inline tests/test_transformations.py:2286]: do we need the second assert? looks to do the same? (`==` = `not !=`?)

**Review by CagtayFabry — APPROVED** [review]
👍

File: tests/test_transformations.py (hunk 11 above)
Comment (vhirtham) [reply tests/test_transformations.py:2286]: Well, technically, one can implement a separate function for `!=` and it doesn't hurt to double-check I guess

File: tests/test_transformations.py (hunk 10 above)
Comment (vhirtham) [reply tests/test_transformations.py:2255]: indeed, it does! 😆

File: tests/test_transformations.py (hunk 9 above)
Comment (vhirtham) [reply tests/test_transformations.py]: Implemented it in a slightly modified version

**Pull request merged**